[Excel] (Custom Functions) Add naming article (#743)
OfficeDev/office-js-docs-pr: docs/excel/custom-functions-naming.md

---
ms.date: 02/08/2019
description: Learn requirements for Excel custom functions' names and avoid common naming pitfalls.
title: Naming guidelines for custom functions in Excel (preview)
localization_priority: Normal
---
# Naming guidelines

> A custom function is identified by an `id` and `name` property in the JSON metadata file. The function `id` is used to uniquely identify custom functions in your JavaScript code. The function `name` is used as the display name that appears to a user in Excel. A function name can differ from the function ID, such as for localization purposes. But in general it should stay the same as the ID if there is no compelling reason for them to differ.

Reviewer: ```suggestion
A custom function is identified by an **id** and **name** property in the JSON metadata file. The function id is used to uniquely identify custom functions in your JavaScript code. The function name is used as the display name that appears to a user in Excel. A function name can differ from the function ID, such as for localization purposes. But in general it should stay the same as the ID if there is no compelling reason for them to differ.
```

Reviewer: I know we often use code comment formatting around this, but I just noticed in this case it reads oddly. When you see function `name` it almost reads as though `name` is a JavaScript function, because it appears as code. But it is actually a property in JSON. What do you think of just bold on first case then use function name afterwards? Does this work?

Reviewer 2: I agree with David. The property should be bold not code comment formatting. That's actually an MSTP thing!

Reviewer 2: Except I'd continue to bold the property each time it is used.

Function names and function IDs share some common requirements:

- They must only use alphanumeric characters (including Unicode), the numbers zero through nine, underscores, and periods.

- They must start with a letter and have a minimum limit of three characters.

> Most function names for built-in Excel functions appear in uppercase letters (such as `SUM`). As a best practice, we recommend to use uppercase for your custom function names so they appear consistent for users.

Reviewer 2: ```suggestion
Excel uses uppercase letters for built-in function names (such as `SUM`). Therefore, consider using uppercase letters for your custom function names and function IDs as a best practice.
```

Reviewer 2: Hmmmm, on the other hand, if people DO want to use a naming convention like Hungarian, lower camel case is very useful, too. Do we really want to suggest this as a best practice?

Reviewer: @lindalu-MSFT  Yes, all functions in Excel appear to users as uppercase (when you type formulas in a cell), so you'll create a more consistent experience by naming your new custom functions also as uppercase.

Reviewer 2: Real coders will still ignore this and use their naming conventions.

> Most function names for built-in Excel functions appear in uppercase letters (e.g. `SUM`). While not required, it is a best practice to specify your custom function names and function IDs using uppercase letters.

Reviewer: ```suggestion
Most function names for built-in Excel functions appear in uppercase letters (such as `SUM`). As a best practice, we recommend to use uppercase for your custom function names so they appear consistent for users.

> Function names should not be named the same as:

Reviewer 2: ```suggestion
Function names shouldn't be named the same as:
```

> Function names should not be named the same as:

Reviewer: won't this cause some kind of error or just unexpected behavior? We should document just so this doesn't appear as an optional kind of thing.

Author: So as far as I know it actually doesn't cause an error - ID automatically converts to be uppercase and name is something that doesn't care about case. We want people to use uppercase because it matches what we do in Excel, but they don't technically have to. I could take out "While not required"?

Reviewer: got it, this looks good then.

Reviewer 2: I agree with Michelle, it should be a best practices. I would simply say "Most function names for built-in Excel functions appear in uppercase letters (e.g. "SUM"). Therefore, it is a best practice  to use uppercase letters for your custom function names and function IDs."

Reviewer 2: I made this in a "suggestion" elsewhere

- Any cells between A1 to XFD1048576 or any cells between R1C1 to R1048576C16384.

> - Any Excel 4.0 Macro Function (such as `RUN`, `ECHO`).  For a full list of these functions, see [this article](https://www.microsoft.com/en-us/download/details.aspx?id=1465).

Reviewer 2: ```suggestion
- Any Excel 4.0 macro function (such as `RUN`, `ECHO`).  For a full list of these functions, download [Macro Function Help File for Excel 4.0 Macros](https://www.microsoft.com/en-us/download/details.aspx?id=1465).
```

Author: As far as I know, this is the only list that Excel provides is in this downloadable file?

> ## Naming conflicts

Reviewer: I think we also need to document some namespace practices. Such as using your company name, product name, team name (for internal add-ins) etc... Check with Keyur for more specifics.

Author: Ok, I will check with him but for now let's get something out in the wild and we can update with namespace info

Reviewer: ok sounds good.

> If a function name in your add-in matches a name in an existing and installed add-in, you'll see an error #REF! appear in your workbook.

Reviewer 2: "If a function name in your add-in is the same as one in a previously installed add-in, you'll see an error #REF! appear in your workbook."

> To troubleshoot this, please change the name in your add-in and try the function again. If you're testing your add-in in different environments, either uninstall the add-in with the conflicting name or use a different namespace to differentiate your function (e.g. NAMESPACE_NAMEOFFUNCTION).

Reviewer: ```suggestion
To fix a name conflict, change the name in your add-in and try the function again. If you're testing your add-in in different environments, either uninstall the add-in with the conflicting name or use a different namespace to differentiate your function (such as NAMESPACE_NAMEOFFUNCTION).
```

> Also consider how you'd like people to use the functions within your add-in. In many cases, it makes sense to add multiple arguments to a function rather than create multiple functions with the same or similar names.

Reviewer: Would be good to show an example here of "instead of this... do this..."

Reviewer 2: I agree with David, and example or two would be great here. When do you add arguments and how do you name them. When do you NOT add arguments. For an example, I once wrote a custom function that returned the size of artworks in our database: GETARTDIMENSIONS(height, width, length) instead of GETARTHEIGHT, GETARTLENGHT, GETARTWIDTH. 

I also think you should consider adding a few more "best practices" sentences about naming functions and arguments so they describe what they do and the importance of readability. They should be named using natural language so others can understand.
1. They shouldn't be pointless like X1, FE2or XBQNE or overly long either, like ADDNEWIFLEGALDRINKINGAGEANDMOONISFULL. They should tell it like it is i.e. ISLEGALDRINKINGAGE. 
2. Avoid ambiguous abbreviations. Clarity is more important than shortness. i.e. choose INCREMENT(), INCREASE() or INCLUDE(), not just INC(). 
3. If a function is an action, it should include the verb in its name: i.e. GETZIPCODE or SETZIPCODE instead of just ZIPCODE. 
4. You might also add a link to some naming schemes such as how to indicate what the function returns. Prefix with S_ for a string function, B_ for boolean, that sort of thing. (Though allcaps instead of lower camel case makes it hard to use Hungarian!) 
6. Shouldn't the function names should be in English--since programming languages are in English, it makes sense, no?)
7. Choose one name for an action and stick with it, don't use both REMOVEZIP and DELETEADDRESS, select one and stick with it. i.e.  REMOVEZIP and REMOVEADDRESS.
